packages/grid/x-data-grid/src/hooks/features/columns/gridColumnsUtils.ts

```diff
@@ -311,39 +313,64 @@ export const createColumnsState = ({
-        columnVisibilityModel[field] = !columnsStateWithoutColumnVisibilityModel.lookup[field].hide;
-      });
-    } else {
-      const newColumnVisibilityModel = { ...currentColumnVisibilityModel };
-      let hasModelChanged = false;
+    let hasModelChanged = false;
 
+    const newColumnVisibilityModel = { ...currentColumnVisibilityModel };
+    if (isInsideStateInitializer) {
       columnsStateWithPreProcessing.all.forEach((field) => {
-        // If neither the `columnsToUpsert` nor the pre-processors updated the column,
-        // Then we don't want to update the visibility status of the column in the model.
-        if (
-          !columnsToUpsertLookup[field] &&
-          columnsLookupBeforePreProcessing[field] === columnsStateWithPreProcessing.lookup[field]
-        ) {
-          return;
+        newColumnVisibilityModel[field] =
+          !columnsStateWithoutColumnVisibilityModel.lookup[field].hide;
+      });
+    } else if (keepOnlyColumnsToUpsert) {
+      // At this point, `keepOnlyColumnsToUpsert` changes its meaning to keep the columns
+      // passed via `columnToUpsert` + columns added by the pre-processors. We do the following
+      // cleanup because a given column may have been removed from the `columns` prop but it still
+      // exists in the state.
+      Object.keys(newColumnVisibilityModel).forEach((field) => {
+        if (!columnsStateWithPreProcessing.lookup[field]) {
+          delete newColumnVisibilityModel[field];
+          hasModelChanged = true;
         }
+      });
+    }
 
-        const isVisibleBefore = currentColumnVisibilityModel[field] ?? true;
-        const isVisibleAfter = !columnsStateWithPreProcessing.lookup[field].hide;
-
-        if (isVisibleAfter !== isVisibleBefore) {
-          hasModelChanged = true;
+    columnsStateWithPreProcessing.all.forEach((field) => {
+      // If neither the `columnsToUpsert` nor the pre-processors updated the column,
+      // Then we don't want to update the visibility status of the column in the model.
+      if (
+        !columnsToUpsertLookup[field] &&
+        columnsLookupBeforePreProcessing[field] === columnsStateWithPreProcessing.lookup[field]
+      ) {
+        return;
+      }
 
-          newColumnVisibilityModel[field] = isVisibleAfter;
+      // We always assume that a column not in the model is visible by default. However, there's an
+      // edge case where the column is not in the model but it also doesn't exist in the `columns`
+      // prop, meaning that the column is being added. In that case, we assume that the column was
+      // not visible before for it be added to the model.
+      let isVisibleBefore = currentColumnVisibilityModel[field];
+      if (isVisibleBefore === undefined) {
+        if (isInsideStateInitializer) {
+          isVisibleBefore = true;
+        } else {
+          const currentState = gridColumnsSelector(apiRef.current.state);
+          isVisibleBefore = currentState.all.includes(field);
```

Comment:
This change was because https://github.com/mui/mui-x/blob/1922f48f0f230cb48bb5915bde0eaec1d0ae8388/packages/grid/x-data-grid/src/tests/layout.DataGrid.test.tsx#L602 was failing.

Reply:
Can we avoid the `O(n²)` complexity here by using the lookup ?